dmriprep/utils/tests/test_vectors.py

```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+import os
+import pytest
+import numpy as np
+from pathlib import Path
+from dmriprep.utils import vectors
+
+
+def test_make_and_save_gradient_table():
+    """Test for gradient table generation and saving from bval/bvec."""
+    from dipy.data import fetch_sherbrooke_3shell
+    from dipy.io import read_bvals_bvecs
+    from os.path import expanduser, join
+    from dipy.core.gradients import GradientTable
+
+    fetch_sherbrooke_3shell()
+    home = expanduser('~')
+    basedir = join(home, '.dipy', 'sherbrooke_3shell')
+    fdwi = join(basedir, 'HARDI193.nii.gz')
+    fbval = join(basedir, 'HARDI193.bval')
+    fbvec = join(basedir, 'HARDI193.bvec')
+
+    # load bvecs/bvals.
+    bvals, bvecs = read_bvals_bvecs(fbval, fbvec)
+
+    b0_thr = 50
+    # Create vector directory if it does not already exist.
+    vector_dir = str(Path(basedir)/'vectors')
+    os.makedirs(vector_dir, exist_ok=True)
+    gtab_tsv_file = str(Path(vector_dir) / 'gradients.tsv')
+
+    [gtab, b0_ixs] = vectors.make_gradient_table(bvals, bvecs, b0_thr)
+    vectors.write_gradients_tsv(bvecs, bvals, gtab_tsv_file)
+
+    [bvals_read, bvecs_read] = vectors.read_gradients_tsv(gtab_tsv_file)
+
+    assert gtab is not None and gtab.__class__ == GradientTable
+    assert gtab.b0_threshold == 50
+    assert np.sum(gtab.b0s_mask) == len(b0_ixs)
+    assert os.path.isfile(gtab_tsv_file) is True
+    assert np.isclose(bvecs_read.all(), bvecs.all())
+    assert np.isclose(bvals_read.all(), bvals.all())
+    os.remove(gtab_tsv_file)
```

Comment:
Too much boilerplate. I'll send a PR showing you how to use the `tmpdir` fixture so that tests run in temporal folders.

Reply:
> I think this is almost there, only the test file remaining. I'll write to your branch directly this time since I believe you won't mind and it will be much faster for iterating.
> 
> Then, I'll squash some commits before asking others to review (finalizing the whole work).
> 
> This has been a bit painful process (especially for you - you've done a tone of work). I hope it's been as illustrative as painful too.
> 
> We want the process to be like a skeet shooting game. When we open an issue, that's shouting for another clay pigeon target. Sending a PR is actually shooting. Tests and reviewers are the referees determining whether we hit the target.
> 
> The earlier version of the PR attempted to hit several targets at once. That's harder for the shooter, and for the referees as well. It also requires more tests. The more focused a PR is, the better. Perhaps, I'm a bit stubborn with this because I actually have fallen for this plenty of times (and sometimes I still do!).
> 
> It is tempting to try to solve everything at once but in the long that usually backfires.
> 
> I'll be super-happy to see this work merged in master!

Yeah but I think to some extent for this PR, this volume of work was perhaps a bit inevitable (and the boundaries harder to define) since we are essentially setting the spec and io all at once that will provide the foundation for vector handling throughout the workflow. That is, once we get this the way we want it (hopefully close), this code will get reused again and again, and in different ways, each time a change is made to the dwi. Vector handling is just super broad and a lot of information/switches are unavoidably tethered to it. I think making more piecewise PR's down the road will be wayyy easier. (e.g. eddy only, denoising only, etc.)